packages/dev/core/src/XR/features/WebXRDepthSensing.ts

```diff
@@ -0,0 +1,320 @@
+import { RawTexture } from "../../Materials/Textures/rawTexture";
+import { WebXRFeatureName, WebXRFeaturesManager } from "../webXRFeaturesManager";
+import type { WebXRSessionManager } from "../webXRSessionManager";
+import { WebXRAbstractFeature } from "./WebXRAbstractFeature";
+import { Tools } from "../../Misc/tools";
+import { Texture } from "../../Materials/Textures/texture";
+import { Engine } from "../../Engines/engine";
+import { Observable } from "../../Misc/observable";
+
+/**
+ * Options for Depth Sensing feature
+ */
+export interface IWebXRDepthSensingOptions {
+    /**
+     *  The desired depth sensing usage for the session
+     */
+    usagePreference: XRDepthUsage[];
+    /**
+     * The desired depth sensing data format for the session
+     */
+    dataFormatPreference: XRDepthDataFormat[];
+}
+
+type GetDepthInMetersType = (x: number, y: number) => number;
+
+/**
+ * WebXR Feature for WebXR Depth Sensing Module
+ */
+export class WebXRDepthSensing extends WebXRAbstractFeature {
+    private _cachedDepthInfo?: XRDepthInformation;
+
+    /**
+     * Width of depth data. If depth data is not exist, returns null.
+     */
+    public get width(): number | null {
+        if (!this._cachedDepthInfo) {
+            return null;
+        }
+        return this._cachedDepthInfo.width;
+    }
+
+    /**
+     * Height of depth data. If depth data is not exist, returns null.
+     */
+    public get height(): number | null {
+        if (!this._cachedDepthInfo) {
+            return null;
+        }
+        return this._cachedDepthInfo.height;
+    }
+
+    /**
+     * Scale factor by which the raw depth values must be multiplied in order to get the depths in meters.
+     */
+    public get rawValueToMeters(): number | null {
+        if (!this._cachedDepthInfo) {
+            return null;
+        }
+
+        return this._cachedDepthInfo.rawValueToMeters;
+    }
+
+    /**
+     * An XRRigidTransform that needs to be applied when indexing into the depth buffer.
+     */
+    public get normDepthBufferFromNormView(): XRRigidTransform | null {
+        if (!this._cachedDepthInfo) {
+            return null;
+        }
+
+        return this._cachedDepthInfo.normDepthBufferFromNormView;
+    }
+
+    /**
+     * Describes which depth-sensing usage ("cpu-optimized" or "gpu-optimized" is used.
+     *
+     */
+    public get depthUsage(): XRDepthUsage {
+        return this._xrSessionManager.session.depthUsage;
+    }
+
+    /**
+     * Describes which depth sensing data format ("luminance-alpha" or "float32") is used.
+     */
+    public get depthDataFormat(): XRDepthDataFormat {
+        return this._xrSessionManager.session.depthDataFormat;
+    }
+
+    /**
+     * Latest cached WebGLTexture which containing depth buffer information.
+     * This can be used when the depth usage is gpu-optimized.
+     */
+    public get latestWebGLTexture(): WebGLTexture | null {

```

Comment:
We do our best to return types that are more helpful to the framework users. It will be great to return here an InternalTexture, configured to already consume this WebGLTexture. You can see an example here - https://github.com/BabylonJS/Babylon.js/blob/master/packages/dev/core/src/XR/features/WebXRLightEstimation.ts#L215 . Though a cube texture, it is roughly the same concept. An internalTexture can then be used with other components of Babylon and the dev doesn't need to create it on their own.

Reply:
Just as a note - this is done so we can stay back-compat in case webxr will ever support webgpu

Reply:
Thank you for your review!
> this is done so we can stay back-compat in case webxr will ever support webgpu
I see, I'll change it to returning InternalTexture.

Reply:
I've resolve this by converting cached WebGLTexture when this getter property called. :+1:

```ts
    public get latestInternalTexture(): Nullable<InternalTexture> {
        if (!this._cachedWebGLTexture) {
            return null;
        }

        const engine = this._xrSessionManager.scene.getEngine();
        const internalTexture = new InternalTexture(engine, InternalTextureSource.Unknown);
        internalTexture.isCube = false;
        internalTexture.invertY = false;
        internalTexture._useSRGBBuffer = false;
        internalTexture.format = this.depthDataFormat === "ushort" ? Constants.TEXTUREFORMAT_LUMINANCE_ALPHA : Constants.TEXTUREFORMAT_RGBA;
        internalTexture.generateMipMaps = false;
        internalTexture.type = this.depthDataFormat === "ushort" ? Constants.TEXTURETYPE_UNSIGNED_SHORT : Constants.TEXTURETYPE_FLOAT;
        internalTexture.samplingMode = Constants.TEXTURE_NEAREST_LINEAR;
        internalTexture.width = this.width ?? 0;
        internalTexture.height = this.height ?? 0;
        internalTexture._cachedWrapU = Constants.TEXTURE_WRAP_ADDRESSMODE;
        internalTexture._cachedWrapV = Constants.TEXTURE_WRAP_ADDRESSMODE;
        internalTexture._hardwareTexture = new WebGLHardwareTexture(this._cachedWebGLTexture, engine._gl);

        return internalTexture;
    }
```